### PR TITLE
GPU Memory workaround work

### PR DESCRIPTION
**Pull request by dkazanc**
from @team-gpu :

This adds memory estimation functions and decorators for the methods, including tests. This is documented in the doc/decorators.rst file as well as docstrings.

What is still missing:
- separate tests for decorators / memory estimation from the functional tests (example done in test_alignment.py)
- some methods are still missing these memory estimator tests. Adding them will reveal mistakes in the functions themselves
- possibly adapting the estimator functions to also return the output datatype, as that might change depending on inputs and parameters
- documentation cleanup

### CODE REVIEW

File: httomolib/recon/algorithm.py [hunk 1]
```diff
@@ -65,92 +90,196 @@ def reconstruct_tomobar(
...
+    # a guess for astra toolbox memory usage for projection/backprojection
+    astra_size = 0.5*(x_rec+data_out)
+   
+    total_mem = int(data_out + x_rec + d + r + Ad + s + astra_size)
+    slices_max = available_memory // total_mem
+    return (slices_max, float32())
+## %%%%%%%%%%%%%%%%%%%%%%% CGLS reconstruction %%%%%%%%%%%%%%%%%%%%%%%%%%%%  ##
+@method_sino(_calc_max_slices_CGLS)
+@nvtx.annotate()
+def CGLS_rec(
```
Comment (namannimmo10) [inline httomolib/recon/algorithm.py:219]: add this to `__all__`?

File: tests/test_recon/rotation_cpu_reference.py [hunk 2]
```diff
@@ -36,7 +37,7 @@ def find_center_360_numpy(
     else:
         cor = ncol - overlap / 2.0 - 1.0
 
-    return np.float32(cor), np.float32(overlap), side, np.float32(overlap_position)
+    return float(cor), float(overlap), side, float(overlap_position)
```
Comment (namannimmo10) [inline tests/test_recon/rotation_cpu_reference.py:40]: Note that `float` in Python is equivalent to `np.float64`, is that what you want?

File: tests/test_recon/test_algorithm.py [hunk 3]
```diff
@@ -28,31 +32,223 @@ def test_reconstruct_tomobar_device_1(data, flats, darks, ensure_clean_memory):
...
+    objrecon_size = data.shape[2]    
+    cp.get_default_memory_pool().free_all_blocks()
+    cache = cp.fft.config.get_plan_cache()
+    cache.clear()
 
+    objrecon_size = data.shape[2]
+    hook = MaxMemoryHook(data.size * data.itemsize)
+    with hook:
+        recon_data = CGLS_rec(
+            data,
```
Comment (namannimmo10) [inline tests/test_recon/test_algorithm.py:233]: here we are passing `data` without normalizing it, we might get the same issue with `NaN`s before...

Comment (dkazanc) [reply tests/test_recon/test_algorithm.py:233]: for `MaxMemoryHook` we are not interested in the result, just the memory usage, so it should be ok here. We normalise for non-hook methods.

File: tests/test_recon/rotation_cpu_reference.py (hunk 2 above)
Comment (dkazanc) [reply tests/test_recon/rotation_cpu_reference.py:40]: it is a scalar, but probably keeping the 32 bit consistency makes sense

**Pull request merged**